RecoHI/HiEgammaAlgos/src/CxCalculator.cc

```diff
@@ -0,0 +1,655 @@
+// ROOT includes
+#include <Math/VectorUtil.h>
+
+#include "RecoHI/HiEgammaAlgos/interface/CxCalculator.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+#include "Geometry/CaloGeometry/interface/CaloSubdetectorGeometry.h"
+
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/EgammaReco/interface/BasicCluster.h"
+#include "DataFormats/EgammaReco/interface/SuperCluster.h"
+#include "DataFormats/Math/interface/Vector3D.h"
+
+
+using namespace edm;
+using namespace reco;
+using namespace std;
+using namespace ROOT::Math::VectorUtil;
+
+#define PI 3.141592653589793238462643383279502884197169399375105820974945
+
+
+double CxCalculator::getJurassicArea( double r1, double r2, double width) {
+
+   float theta1 = asin( width / r1);
+   float theta2 = asin( width / r2);
+   float theA   = sqrt ( r1*r1 + r2*r2 - 2 * r1 * r2 * cos ( theta1 - theta2) );
+   float area1 =  0.5 * r1*r1 * ( 3.141592 - 2 * theta1 )   ;
+   float area2 =  0.5 * r2*r2 * ( 3.141592 - 2 * theta2 )   ;
+   float area3 =  width * theA;
+   float finalArea = 2 * ( area1 - area2 - area3);
+   return finalArea;
+}
+
+CxCalculator::CxCalculator(const edm::Event &iEvent, const edm::EventSetup &iSetup, edm::Handle<BasicClusterCollection> pEBclusters, edm::Handle<BasicClusterCollection> pEEclusters)
+{
+  if(pEBclusters.isValid())
+     fEBclusters_ = pEBclusters.product();
+   else
+     fEBclusters_ = NULL;
+
+   if(pEEclusters.isValid())
+     fEEclusters_ = pEEclusters.product();
+   else
+     fEEclusters_ = NULL;
+
+   ESHandle<CaloGeometry> geometryHandle;
+   iSetup.get<CaloGeometryRecord>().get(geometryHandle);
+   if(geometryHandle.isValid())
+     geometry_ = geometryHandle.product();
+   else
+     geometry_ = NULL;
+}
+
+double CxCalculator::getCx(const reco::SuperClusterRef cluster, double x, double threshold)
+{
+   using namespace edm;
+   using namespace reco;
+
+   if(!fEBclusters_) {
+//       LogError("CxCalculator") << "Error! Can't get EBclusters for event.";
+      return -100;
+   }
+
+   if(!fEEclusters_) {
+//       LogError("CxCalculator") << "Error! Can't get EEclusters for event.";
+      return -100;
+   }
+
+   math::XYZVector SClusPoint(cluster->position().x(),
+                          cluster->position().y(),
+                          cluster->position().z());
+
+   double TotalEt = 0;
+
+   TotalEt = - cluster->rawEnergy()/cosh(cluster->eta());
+
+   // Loop over barrel basic clusters
+   for(BasicClusterCollection::const_iterator iclu = fEBclusters_->begin();
+       iclu != fEBclusters_->end(); ++iclu) {
+      const BasicCluster *clu = &(*iclu);
+      math::XYZVector ClusPoint(clu->x(),clu->y(),clu->z());
+      double eta = ClusPoint.eta();
+
+      double dR = ROOT::Math::VectorUtil::DeltaR(ClusPoint,SClusPoint);
```

Comment:
use ``` reco::deltaR2(*clu, cluster)```:
- faster
- no need for extra code to extract points
- no need for unneeded sqrt (more expensive than x*x*0.01 below